packages/lit-renderer/src/lit-renderer.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * @license
+ * Copyright (c) 2016 - 2022 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+import { nothing, render } from 'lit';
+import { AsyncDirective } from 'lit/async-directive.js';
+import { PartType } from 'lit/directive.js';
+
+// A symbol indicating that the directive hasn't been initialized.
+const VALUE_NOT_INITIALIZED = Symbol('valueNotInitialized');
+
+export class LitRendererDirective extends AsyncDirective {
+  /** @protected */
+  previousValue = VALUE_NOT_INITIALIZED;
+
+  constructor(part) {
+    super(part);
+    if (part.type !== PartType.ELEMENT) {
+      throw new Error(`The \`${this.name}\` directive must be bound to an element.`);
+    }
+  }
+
+  /** @override */
+  update(part, [renderer, value]) {
+    if (!this.hasChanged(value)) {
+      return nothing;
+    }
+
+    this.host = part.options && part.options.host;
+    this.element = part.element;
+    this.renderer = renderer;
+
+    const firstRender = this.previousValue === VALUE_NOT_INITIALIZED;
+    if (firstRender) {
+      this.addRenderer();
+    } else {
+      this.runRenderer();
+    }
+
+    // Copy the value if it is an array in order to keep it
+    // from possible outside mutations.
+    this.previousValue = Array.isArray(value) ? [...value] : value;
+
+    return nothing;
+  }
+
+  /** @override */
+  reconnected() {
+    this.addRenderer();
+  }
+
+  /** @override */
+  disconnected() {
+    this.removeRenderer();
+  }
+
+  /** @abstract */
+  get name() {
+    throw new Error('The `name` getter must be implemented.');
+  }
+
+  /** @abstract */
+  addRenderer() {
+    throw new Error('The `addRenderer` method must be implemented.');
+  }
+
+  /** @abstract */
+  runRenderer() {
+    throw new Error('The `runRenderer` method must be implemented.');
+  }
+
+  /** @abstract */
+  removeRenderer() {
+    throw new Error('The `removeRenderer` method must be implemented.');
+  }
+
+  /** @protected */
+  renderRenderer(container, ...args) {
+    const templateResult = this.renderer.call(this.host, ...args);
```

Comment:
Consider using a property that would default to `renderer`, to make it possible to specify `headerRenderer` and `footerRenderer` for grid column and dialog components.

Reply:
Not sure if I got the idea. The reason I extracted `renderRenderer` is to get rid of the necessity to call the boilerplate code related to passing `host` to the Lit method `render` in every specific directive. Assigning the renderer to a property is still the responsibility of specific directives. Look at the example:

```js
class SelectRendererDirective extends LitRendererDirective() {
  // ...

  addRenderer() {
    this.renderer = (root, model) => this.renderRenderer(root, model);
  }

  // ...
}
``` 

Are you suggesting specifying the name of a component's renderer property in a directive's generic property like `rendererProperty`?

Reply:
Sounds good 👍  Let's keep it as is then.